File: server/src/main/java/com/kernith/easyinvoice/data/model/state/invoice/InvoiceStateFactory.java
package com.kernith.easyinvoice.data.model.state.invoice;

import com.kernith.easyinvoice.data.model.Invoice;
import com.kernith.easyinvoice.data.model.InvoiceStatus;

/**
 * Factory that maps invoice status to the appropriate state handler.
 */
public class InvoiceStateFactory {
    private InvoiceStateFactory() {}

    public static InvoiceState from(Invoice invoice) {
        if (invoice == null) {
            throw new IllegalArgumentException("Invoice is required");
        }
        return from(invoice.getStatus());
    }

    public static InvoiceState from(InvoiceStatus status) {
        return switch (status) {
            case DRAFT -> new DraftInvoiceState();
            case ISSUED -> new IssuedInvoiceState();
            case PAID -> new PaidInvoiceState();
            case OVERDUE -> new OverdueInvoiceState();
            case ARCHIVED -> new ArchivedInvoiceState();
        };
    }
}

File: client/src/core/state/invoice-state-factory.ts
import { InvoiceStatus } from "./invoice-status";
import { InvoiceState } from "./invoice-state";
import {
  ArchivedInvoiceState,
  DraftInvoiceState,
  IssuedInvoiceState,
  OverdueInvoiceState,
  PaidInvoiceState,
} from "./invoice-states";

/** Factory that maps invoice status to a state handler. */
export class InvoiceStateFactory {
  private constructor() {}

  /**
   * @param status current invoice status
   */
  static from(status: InvoiceStatus): InvoiceState {
    switch (status) {
      case InvoiceStatus.DRAFT:
        return new DraftInvoiceState();
      case InvoiceStatus.ISSUED:
        return new IssuedInvoiceState();
      case InvoiceStatus.PAID:
        return new PaidInvoiceState();
      case InvoiceStatus.OVERDUE:
        return new OverdueInvoiceState();
      case InvoiceStatus.ARCHIVED:
        return new ArchivedInvoiceState();
      default:
        return new DraftInvoiceState();
    }
  }
}
